### PR TITLE
optimizer: add PushFilterDownProjectRule

### PR DESCRIPTION
**Pull request by zzl200012**

### CODE REVIEW

**Review by jievince — COMMENTED** [review]
Excellent job! Can you add some tck tests?

File: src/optimizer/rule/PushFilterDownProjectRule.cpp [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+    auto filterGroupNode = matched.node;
+    auto oldFilterNode = filterGroupNode->node();
+    auto projGroupNode = matched.dependencies.front().node;
+    auto oldProjNode = projGroupNode->node();
+
+    auto newFilterNode = static_cast<graph::Filter*>(oldFilterNode->clone());
+    auto newProjNode = static_cast<graph::Project*>(oldProjNode->clone());
+
+    // Exchange planNode
+    newProjNode->setOutputVar(oldFilterNode->outputVar());
```
Comment (czpmango) [inline src/optimizer/rule/PushFilterDownProjectRule.cpp:46]: Non-VariablePropertyExpressions may need to be rewritten. 


Before:
`Project : a,abs(b) AS b,c+1 AS c => Filter: a>1 and b-1>3 and c>3`
After:
`Filter : a>1 and abs(b)-1>3 and c+1>3  => Project: a, abs(b) AS b, c+1 AS c`

File: tests/tck/features/optimizer/PushFilterDownProjectRule.feature [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+      """
+      MATCH (a:player)--(b)--(c)
+      WITH
+        a AS a,
+        b AS b,
+        c AS c
+      WHERE a.age < 25
+      AND b.age < 25
+      AND c.age < 25
+      RETURN a
```
Comment (czpmango) [inline tests/tck/features/optimizer/PushFilterDownProjectRule.feature:21]: format

File: tests/tck/features/optimizer/PushFilterDownProjectRule.feature [hunk 3]
```diff
@@ -0,0 +1,47 @@
+# Copyright (c) 2021 vesoft inc. All rights reserved.
+#
+# This source code is licensed under Apache 2.0 License,
+# attached with Common Clause Condition 1.0, found in the LICENSES directory.
+Feature: Push Filter down Project rule
+
+  Background:
+    Given a graph with space named "nba"
+
+  Scenario: push filter down Project
```
Comment (czpmango) [inline tests/tck/features/optimizer/PushFilterDownProjectRule.feature:10]: Add this case:
```
MATCH (a:player)--(b)--(c) 
WITH a, b, c.age + 1 AS cage 
WHERE a.name=="Tim Duncan" AND b.age>40 
RETURN a,b,cage
```

File: src/optimizer/rule/PushFilterDownProjectRule.cpp [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+    auto oldProjNode = projGroupNode->node();
+
+    auto newFilterNode = static_cast<graph::Filter*>(oldFilterNode->clone());
+    auto newProjNode = static_cast<graph::Project*>(oldProjNode->clone());
+    const auto condition = newFilterNode->condition();
+
+    auto varProps = graph::ExpressionUtils::collectAll(condition, {Expression::Kind::kVarProperty});
+    if (varProps.size() == 0) {
+        return TransformResult::noTransform();
+    }
```
Comment (czpmango) [inline src/optimizer/rule/PushFilterDownProjectRule.cpp:49]: varProps.empty()

**Review by czpmango — APPROVED** [review]

**Review by jievince — APPROVED** [review]

**Review by yixinglu — APPROVED** [review]

**Comment by jamieliu1023** [comment]
@zzl200012 Thanks for your contribution to the Nebula Graph project! This is Jamie with Nebula. I would like to email you your Nebula Contributor certificate and ship you a mug to mark this special moment! Could you please kindly let me know your email address? You can reach me via jamie.liu(at)vesoft.com. I am looking forward to your email. Thanks again!

**Pull request merged**